posts/multiple-inheritance-crowdsales.md

```diff
@@ -0,0 +1,281 @@
+Multiple inheritance is hard, and it can get very confusing if you abuse it.
+Some languages don't support it at all; but it is becoming a very common
+pattern on Solidity. Thus, one of the first things to do when starting to
+learn Solidity is to fully understand how it handles multiple inheritance.
+
+The problem, from the point of view of the programmer, is to understand the
+order of the calls when a contract has multiple parents. Let's say we have a
+base contract `A`, with a function named `f`:
+
+```
+// in A.
+function f() {
+  somethingA();
+}
+```
+
+Then we have three contracts `B`, `C`, and `D`, all of them overriding `f`. `B`
+does something like this:
+
+```
+// in B.
+function f() {
+  somethingB();
+  super.f();
+}
+```
+
+`C` does something like this:
+
+```
+// in C.
+function f() {
+  somethingC();
+  super.f();  
+}
+```
+
+And `D` does something like this:
+
+```
+// in D.
+function f() {
+  somethingD();
+  super.f();  
+}
+```
+
+What happens when you call `D.f`?
+
+This is called the diamond problem, because we end up with a diamond-shaped
+inheritance diagram:
+
```

Comment:
It's confusing to have this diagram show up without the class declarations stipulating what the inheritance relations are.. I understand you want to discuss this later, but I would at least put one as an example here, like `contract D is B, C`, and also `contract B is A` and `contract C is A`, otherwise there is no clue as to where the diagram is coming from.. (To me at least, the "... overriding `f`" is not enough.)